### PR TITLE
Better controller comm 15

### LINKED ISSUE #15: improve controller communication
request to controller should be made on the event loop instead of background thread
enable periodic polling of available services

### PR DESCRIPTION
**Pull request by ekoby**
Closes #15 

this changes SDK-to-EdgeController communication to be fully async and executed on the event loop provided to SDK.

Services and Network Sessions are no longer preloaded during NF_init() but instead created on demand.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,7 +20,8 @@ if (WIN32)
     message("WIN32 build. Creating: ${CMAKE_BINARY_DIR}/cmake_install/ziti-sdk-${PROJECT_VERSION}")
     file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/cmake_install/ziti-sdk-${PROJECT_VERSION})
     set(CMAKE_INSTALL_PREFIX ${CMAKE_BINARY_DIR}/cmake_install/ziti-sdk-${PROJECT_VERSION})
-    add_subdirectory(windows)
+    # TODO 
+    # add_subdirectory(windows)
```
Comment (dovholuknf) [inline CMakeLists.txt]: perhaps make an issue for this?

Comment (ekoby) [reply CMakeLists.txt]: I added general issue for build a shared lib already #24

File: inc_internal/ziti_ctrl.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
+/*
+Copyright (c) 2019 Netfoundry, Inc.
```
Comment (dovholuknf) [inline inc_internal/ziti_ctrl.h]: moving a file is a 'touch'? needs 2019-2020?

Comment (ekoby) [reply inc_internal/ziti_ctrl.h]: missed it in update. fixed

File: library/controller.h [hunk 3]
```diff
@@ -24,7 +24,7 @@ typedef int (*ctrl_req)(struct nf_ctx *, uv_os_sock_t sock, tls_engine *ssl);
 int ziti_ctrl_process(nf_context ctx, ...);
 
 int ziti_ctrl_version(nf_context ctx, uv_os_sock_t, tls_engine *);
-int ziti_ctrl_login(nf_context ctx, uv_os_sock_t, tls_engine *);
+int ziti_ctrl_login_old(nf_context ctx, uv_os_sock_t, tls_engine *);
```
Comment (dovholuknf) [inline library/controller.h]: needed?

**Comment by dovholuknf** [comment]
There's a mix of free vs FREE used - do you want to use FREE everywhere?

File: inc_internal/ziti_ctrl.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
+/*
+Copyright (c) 2019-2020Netfoundry, Inc.
```
Comment (dovholuknf) [inline inc_internal/ziti_ctrl.h]: almost: 2020Netfoundry

File: library/ziti.c [hunk 5]
```diff
@@ -307,3 +318,33 @@ extern int NF_listen(nf_connection serv_conn, const char *service, nf_listen_cb
...
+    if (err != NULL) {
+        ZITI_LOG(ERROR, "failed to get controller version from %s:%s %s(%s)",
+                 ctrl->client.host, ctrl->client.port, err->code, err->message);
+        free_ziti_error(err);
+    }
+    else {
+        ZITI_LOG(INFO, "connected to controller %s:%s version %s(%s %s)",
+                 ctrl->client.host, ctrl->client.port, v->version, v->revision, v->build_date);
+        free_ctrl_version(v);
+    }
```
Comment (dovholuknf) [inline library/ziti.c]: you aren't free'ing the err here like in other functions - intentional?

Comment (dovholuknf) [reply library/ziti.c]: lol - the fucntion above does it - just didn't notice somehow!

File: library/ziti_ctrl.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        }
+    }
+}
+
+static void ctrl_default_cb (void *s, ziti_error *e, struct ctrl_resp *resp) {
+    if (resp->resp_cb) {
+        resp->resp_cb(s, e, resp->ctx);
+    }
+
+    free(resp);
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: what will free the resp->ctx and resp->body etc? that all handled elsewhere?

File: library/ziti_ctrl.c [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+static void ctrl_login_cb(ziti_session *s, ziti_error *e, struct ctrl_resp *resp) {
+    resp->ctrl->session = strdup(s->token);
+    um_http_header(&resp->ctrl->client, "zt-session", s->token);
+    ctrl_default_cb(s, e, resp);
+}
+
+static void ctrl_logout_cb(void *s, ziti_error *e, struct ctrl_resp *resp) {
+    FREE(resp->ctrl->session);
+    um_http_header(&resp->ctrl->client, "zt-session", NULL);
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: there's no way to just remove a header? or is setting a header to NULL actually removing it from the client?

File: library/ziti_ctrl.c [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    ctrl_default_cb(s, e, resp);
+}
+
+static void ctrl_service_cb(ziti_service **services, ziti_error *e, struct ctrl_resp *resp) {
+    ziti_service *s = services != NULL ? services[0] : NULL;
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: if services isn't null there's guaranteed to be at least one service? if all services are removed from the identity at the controller will services get set to NULL?

File: library/ziti_ctrl.c [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        resp->received += len;
+    } else if (len == UV_EOF) {
+        const char* data;
+        int data_len;
+        void *resp_obj = NULL;
+        ziti_error *err = NULL;
+
+        if (resp->status > 299) {
+            mjson_find(resp->body, resp->received, "$.error", (const char **) &data, &data_len);
+            err = parse_ziti_error(data, data_len);
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: 300's are 'errors'?

File: library/ziti_ctrl.c [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    resp->ctrl_cb = ctrl_service_cb;
+
+    req->data = resp;
+}
+
+void ziti_ctrl_get_net_session(
+        ziti_controller *ctrl, ziti_service *service,
+        void (*cb)(ziti_net_session *, ziti_error*, void*), void* ctx) {
+
+    char *content = malloc(128);
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: anything free'ing content?

File: library/ziti_ctrl.c [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    size_t len = (size_t) sprintf(content, "{\"serviceId\":\"%s\",\"hosting\":%s}",
+                                      service->id, service->hostable ? "true" : "false");
+
+    um_http_req_t *req = um_http_req(&ctrl->client, "POST", "/sessions");
+    req->resp_cb = ctrl_resp_cb;
+    req->body_cb = ctrl_body_cb;
+    um_http_req_header(req, "Content-Type", "application/json");
+    um_http_req_data(req, content, len, free_body_cb);
```
Comment (dovholuknf) [inline library/ziti_ctrl.c]: can len be > 128?

File: library/controller.h (hunk 3 above)
Comment (ekoby) [reply library/controller.h]: file removed

File: inc_internal/ziti_ctrl.h (hunk 4 above)
Comment (ekoby) [reply inc_internal/ziti_ctrl.h]: yep

File: library/ziti_ctrl.c (hunk 7 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: setting to NULL === remove

File: library/ziti_ctrl.c (hunk 9 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: we don't expect them since controller does not do redirects

File: library/ziti_ctrl.c (hunk 8 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: services is  NULL terminated array, if we get empty array then services[0] would be NULL

File: library/ziti_ctrl.c (hunk 11 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: we're formatting request above, so why would it be >128?

File: library/ziti_ctrl.c (hunk 6 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: `resp->ctx` is client owned data, `resp->body` is freed after it's parsed into an API object. line 130

File: library/ziti_ctrl.c (hunk 10 above)
Comment (ekoby) [reply library/ziti_ctrl.c]: line 266, freed in the callback

Comment (dovholuknf) [reply library/ziti_ctrl.c]: Ah, content == body... Thanks

**Pull request merged**